### PR TITLE
Create a documentation about PR

### PR DESCRIPTION
**Pull request by bmarty**
Opened for comment!

[rendered](https://github.com/vector-im/element-android/blob/0dbd6797f7800776edf16cd455a13ee9ed66853d/docs/pull_request.md)

### CODE REVIEW

File: docs/pull_request.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+Draft PR can be created when the submitter does not expect the PR to be reviewed and merged yet. It can be useful to publicly show the work, or to do a self-review first.
+
+Draft PR can also be created when it depends on other un-merged PR.
+
+In any case, it is better to explicitly declare in the description why the PR is a draft PR.
+
+##### Assigned to reviewer?
+
+For the moment we do not directly assign PR to ask for a review. In the future we could have some automation to do that. Anyway the core team should have a look to every PR, even when not digging into details. More reviewers, more eys on the code change, more valuable feedback!
```
Comment (manuroe) [inline docs/pull_request.md:33]:
```suggestion
For the moment we do not directly assign PR to ask for a review. In the future we could have some automation to do that. Anyway the core team should have a look to every PR, even when not digging into details. More reviewers, more eyes on the code change, more valuable feedback!
```

File: docs/pull_request.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+1. Code should follow the guidelines
+2. Code should be formatted correctly
+3. XML attribute must be sorted
+4. New code is added at the correct location
+5. New classes are added to the correct location
+6. Naming is correct. Naming is really important, it's considered part of the documentation
+7. Architecture is followed. For instance, the logic is in the ViewModel and not in the Fragment
+8. There is at least one file for the changelog. Exception if the PR fixes something which has not been released yet.
+9. PR includes test. allScreensTest when applicable, and unit tests
```
Comment (manuroe) [inline docs/pull_request.md:110]:
```suggestion
9. PR includes tests. allScreensTest when applicable, and unit tests
```

😇

File: docs/pull_request.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+10. Avoid over complicating things. Keep it simple (KISS)!
+11. PR contains only the expected change. Sometimes, the diff is showing changes that are already on develop. This is not good, submitter has to fix that up.
+
+##### Check the commit
+
+Commit message must be short, one line and valuable. "WIP" is not a good commit message.
+
+Also commit history should be nice. Having commits like "Adding temporary code" then later "Removing temporary code" is not good. The branch has to be rebased and those commit have to be dropped.
+
+PR merger could decide to squash and merge if commit history is not code.
```
Comment (manuroe) [inline docs/pull_request.md:120]:
```suggestion
PR merger could decide to squash and merge if commit history is not good.
```
?

File: docs/pull_request.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+Example:
+> Closes #1
+
+Note that you have to repeat the keyword in case of a list of issue
+
+> Closes #1, Closes #2, etc.
+
+When PR will be merged, such referenced issue will be automatically closed.
+It up to the person who has merged the PR to go to the (closed) issue(s) and to add a comment to inform in which version the issue fix will be available. Use the current version of develop branch.
```
Comment (manuroe) [inline docs/pull_request.md:146]:
```suggestion
It is up to the person who has merged the PR to go to the (closed) issue(s) and to add a comment to inform in which version the issue fix will be available. Use the current version of develop branch.
```

File: docs/pull_request.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+> Closed in Element Android v1.x.y
+
+### Merge conflict
+
+It's up to the submitter to handle merge conflict. Sometimes, they can be fixed directly from GitHub, sometimes this is not possible. The branch can be rebased on develop, or develop can be merged on the branch, it's up to the submitter to decide what is best.
```
Comment (manuroe) [inline docs/pull_request.md:152]: We could add that GH actions do not run if there are merge conflicts.

File: docs/pull_request.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+> Closes #1
+
+Note that you have to repeat the keyword in case of a list of issue
+
+> Closes #1, Closes #2, etc.
+
+When PR will be merged, such referenced issue will be automatically closed.
+It up to the person who has merged the PR to go to the (closed) issue(s) and to add a comment to inform in which version the issue fix will be available. Use the current version of develop branch.
+
+> Closed in Element Android v1.x.y
```
Comment (manuroe) [inline docs/pull_request.md:148]: FTR, we do not have that on that on other projects (EI, EW, Synapse).

File: docs/pull_request.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+##### File change
+
+1. Code should follow the guidelines
+2. Code should be formatted correctly
+3. XML attribute must be sorted
+4. New code is added at the correct location
+5. New classes are added to the correct location
+6. Naming is correct. Naming is really important, it's considered part of the documentation
+7. Architecture is followed. For instance, the logic is in the ViewModel and not in the Fragment
+8. There is at least one file for the changelog. Exception if the PR fixes something which has not been released yet.
```
Comment (ariskotsomitopoulos) [inline docs/pull_request.md:109]: Maybe add a link here explaining how the `changelog` works, .misc .sdk .feature etc

Comment (bmarty) [reply docs/pull_request.md:109]: Yes I agree, as a general feedback, this file lacks lot of links to both internal and external locations.

File: docs/pull_request.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+> Closed in Element Android v1.x.y
+
+### Merge conflict
+
+It's up to the submitter to handle merge conflict. Sometimes, they can be fixed directly from GitHub, sometimes this is not possible. The branch can be rebased on develop, or develop can be merged on the branch, it's up to the submitter to decide what is best.
+
+### When and who can merge PR
+
+PR can be merged by the submitter, if and only if at least one approval from another developer is done. Approval from all people added as reviewer is also a good thing to have. Approval from design team may be mandatory, but is not sufficient to merge a PR.
```
Comment (manuroe) [inline docs/pull_request.md:156]: Should it be "PR **must** be merged by the submitter" ? For submitters who can, those who are maintainer of the repo.
Sometimes submitters forget that they have to commit.

File: docs/pull_request.md (hunk 6 above)
Comment (bmarty) [reply docs/pull_request.md:148]: Yes, but I think it is very valuable for issue submitters and for every one to see in this issue timeline in which version of the app the issue is actually fixed. We could also decide to close issue only once the code is released to avoid misunderstanding, but it's breaking automation (issue closed when merging PR).
I would suggest that other projects should do that too :)

File: docs/pull_request.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+6. Naming is correct. Naming is really important, it's considered part of the documentation
+7. Architecture is followed. For instance, the logic is in the ViewModel and not in the Fragment
+8. There is at least one file for the changelog. Exception if the PR fixes something which has not been released yet.
+9. PR includes test. allScreensTest when applicable, and unit tests
+10. Avoid over complicating things. Keep it simple (KISS)!
+11. PR contains only the expected change. Sometimes, the diff is showing changes that are already on develop. This is not good, submitter has to fix that up.
+
+##### Check the commit
+
+Commit message must be short, one line and valuable. "WIP" is not a good commit message.
```
Comment (stefanceriu) [inline docs/pull_request.md:116]: I wouldn't necessarily say they need to be short. Most git clients are okay with extra details being written on secondary lines.
Ideally they would also contain a link to the issue that it's referring to e.g. `element-android/issues/XXXX`

File: docs/pull_request.md (hunk 8 above)
Comment (bmarty) [reply docs/pull_request.md:156]: PR should be merged by code owners (or project owners) and not submitters from my point of view. Code owners could decide what can be part of the next release or not, and judge the "risk" of merging a PR regarding the time left until the next scheduled release.

File: docs/pull_request.md (hunk 5 above)
Comment (bmarty) [reply docs/pull_request.md:152]: will do, thanks.

File: docs/pull_request.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+## Submitting PR
+
+### Who can submit pull requests?
+
+Basically every one who wants to contribute to the project! But there are some rules to follow.
+
+#### Humans
+
+Developers who are employed (or paid) by Element can directly clone the project and push their branches and create PR.
```
[the remark this answers is not in the document]
Comment (bmarty) [reply docs/pull_request.md:19]: Will update, thanks

**Review by stefanceriu — COMMENTED** [review]
Looks great 👍

File: docs/pull_request.md [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+As an open source project, every one can review each others PR. Of course an approval from internal developer is mandatory for a PR to be merged.
+But comment in PR from the community are always appreciated!
+
+### What to have in mind when reviewing a PR
+
+1. User experience: is the UX and UI correct? You will probably be the second person to test the new thing, the first one is the developer.
+2. Developer experience: does the code look nice and decoupled? No big functions, etc.
+3. Code maintenance. A bit similar to point 2. Tricky code must be documented for instance
+4. Fork consideration. Is configuration of forks will be easy? Some documentation may help in some cases.
```
Comment (bmarty) [inline docs/pull_request.md:90]: Note to myself: add a paragraph about the "Resolve conversation" thing. I do not like it because you have to open the conversation again to see the content, it's a waste of time.

File: docs/pull_request.md [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+Big PR can take time, and there is a risk of future merge conflict.
+
+Feature flag can be used to avoid half implemented feature to be available in the application.
+
+That said, splitting into several PRs should not have the side effect to have more review to do, for instance if some code is added, then finally removed.
+
+##### Avoid unrelated fixing other issue in a big PR
+
+Each PR should focus on a single task. If other issues may be fixed when working in the area of it, it's preferable to open a dedicated PR.
```
Comment (langleyd) [inline docs/pull_request.md:47]: A good example I think is file formatting. If formatting more than a line or two the file formatting can really obscure the intended purpose of code changes in a PR. Opening a PR that just formats a group of related files can make reviewers lives much easier.

File: docs/pull_request.md [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+ - Check the release note. Some existing bugs in Element project may be fixed by the upgrade
+ - Make sure that the CI is happy
+ - If the code does not compile (API break for instance), you have to checkout the branch and push new commits
+ - Do some smoke test, depending of the library which has been upgraded
+
+For some reason dependabot sometimes does not upgrade some dependencies. In this case, and when detected, the upgrade has to be done manually.
+
+##### Gradle wrapper
+
+`Update Gradle Wrapper` is a tool which can create PR tu upgrade our gradle.properties file.
```
Comment (langleyd) [inline docs/pull_request.md:70]:
```suggestion
`Update Gradle Wrapper` is a tool which can create PRs to upgrade our gradle.properties file.
```

File: docs/pull_request.md [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+Review such PR is the same recipe than for PR from Dependabot
+
+##### Sync analytics plan
+
+This tools import any update in the analytics plan. See instruction in the PR itself to handle it.
```
Comment (langleyd) [inline docs/pull_request.md:75]:
```suggestion
This tools imports any update in the analytics plan. See instruction in the PR itself to handle it.
```

File: docs/pull_request.md [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+##### Check the substance
+
+1. Test the changes!
+2. Test the nominal case and the edge cases
+3. Run the sanity test for critical PR
+
+##### Make a dedicated meeting to review the PR
+
+Some times, big PR can be hard to review. Setting up a call with the PR submitter can faster the communication, rather than putting comment and questions as GitHub comment. It has the inconvenient to not make it public.
```
Comment (langleyd) [inline docs/pull_request.md:132]:
```suggestion
Sometimes a big PR can be hard to review. Setting up a call with the PR submitter can speed up the communication, rather than putting comments and questions in GitHub comment. It has the inconvenience of making the discussion non-public, consider including a summary of the main points of the "offline" conversation in the PR.
```

File: docs/pull_request.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply docs/pull_request.md:19]: done

File: docs/pull_request.md (hunk 11 above)
Comment (bmarty) [reply docs/pull_request.md]: done

File: docs/pull_request.md (hunk 12 above)
Comment (bmarty) [reply docs/pull_request.md:47]: I agree; I will add it

File: docs/pull_request.md (hunk 14 above)
Comment (bmarty) [reply docs/pull_request.md:77]: thanks

File: docs/pull_request.md [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+
+In any case, it is better to explicitly declare in the description why the PR is a draft PR.
+
+Also, draft PR should not stay indefinitely in this state. It may be removed if it is the case and the submitter does not update it after a few days.
+
+##### PR Review Assignment
+
+We use automatic assignment for PR reviews. A PR is automatically routed by GitHub to a team member using the round robin algorithm. The process is the following:
+
+- The PR creator assigns the [element-android](https://github.com/orgs/vector-im/teams/element-android) team as a reviewer. They can skip this process and assign directly a specific member if they think they should take a look at it.
```
Comment (ouchadam) [inline docs/pull_request.md:37]: thinking out loud (not a blocker for this PR) 

could we automate this assignment? similar to how we apply the triage project to new issues~
I'm happy to have a go at updating the github actions

File: docs/pull_request.md (hunk 12 above)
Comment (ouchadam) [reply docs/pull_request.md:70]: formatting and noisy renames as separate PRs would be greatly appreciated :pray:

File: docs/pull_request.md [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+## Reviewing PR
+
+### Who can review pull requests?
+
+As an open source project, every one can review each others PR. Of course an approval from internal developer is mandatory for a PR to be merged.
+But comment in PR from the community are always appreciated!
+
+### What to have in mind when reviewing a PR
+
+1. User experience: is the UX and UI correct? You will probably be the second person to test the new thing, the first one is the developer.
```
[the remark this answers is not in the document]
Comment (ouchadam) [reply docs/pull_request.md:112]: thinking out loud :smile: 

Ideally design/product would be involved/signed off before the change is implemented, letting us focus on the edgecases/potential errors in the implementation

It feels a little late to be questioning the core UX/UI after the change has been implemented :thinking: and potentially a little disheartening to find the thing you've built be rejected after investing time on it

File: docs/pull_request.md (hunk 16 above)
Comment (manuroe) [reply docs/pull_request.md:37]: This is a good idea but we still want to be able to assign someone directly. If we change the selected reviewer afterwards, it will create noise.

There is a missing paragraph from the original doc that was used to create this section. I will add a comment.

File: docs/pull_request.md [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+
+Draft PR can be created when the submitter does not expect the PR to be reviewed and merged yet. It can be useful to publicly show the work, or to do a self-review first.
+
+Draft PR can also be created when it depends on other un-merged PR.
+
+In any case, it is better to explicitly declare in the description why the PR is a draft PR.
+
+Also, draft PR should not stay indefinitely in this state. It may be removed if it is the case and the submitter does not update it after a few days.
+
+##### PR Review Assignment
```
Comment (manuroe) [inline docs/pull_request.md:33]:
```suggestion
## PR Review Assignment
```

I think this section deserves higher visibility. It has a huge impact on daily work. We should be able to send people directly to this section and the title must be more visible.

File: docs/pull_request.md [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+
+We use automatic assignment for PR reviews. A PR is automatically routed by GitHub to a team member using the round robin algorithm. The process is the following:
+
+- The PR creator assigns the [element-android](https://github.com/orgs/vector-im/teams/element-android) team as a reviewer. They can skip this process and assign directly a specific member if they think they should take a look at it.
+- GitHub automatically assigns one reviewer. If the chosen reviewer is not available (holiday, etc.), remove them and manually add another reviewer.
+- The reviewer gets a notification to make the review: they review the code following the good practice (see the rest of this document).
+- After making their own review, if they feel not confident enough, they can ask another person for a full review, or they can tag someone within a PR comment to check specific lines.
+
+For PRs coming from the community, the issue wrangler can assign either the team [element-android](https://github.com/orgs/vector-im/teams/element-android) or any member directly.
+
```
Comment (manuroe) [inline docs/pull_request.md:43]: What about this paragraph from the original doc?
```
If a PR depends on another PR, assign the same reviewer, even if they are from different repositories. For example, related matrix-ios-sdk and element-ios PRs should be reviewed by the same person. Run the automatic assignment on the first PR and use the selected reviewer for the second PR.
```

File: docs/pull_request.md [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+As an open source project, every one can review each others PR. Of course an approval from internal developer is mandatory for a PR to be merged.
+But comment in PR from the community are always appreciated!
+
+### What to have in mind when reviewing a PR
+
+1. User experience: is the UX and UI correct? You will probably be the second person to test the new thing, the first one is the developer.
+2. Developer experience: does the code look nice and decoupled? No big functions, new classes added to the right module, etc.
+3. Code maintenance. A bit similar to point 2. Tricky code must be documented for instance
+4. Fork consideration. Will configuration of forks be easy? Some documentation may help in some cases.
+5. We are building long term products. "Quick and dirty" code must be avoided.
```
Comment (ouchadam) [inline docs/pull_request.md:116]: do you think it's worth including some examples? I've noticed a trend in community PRs commenting out existing logic rather than deleting the no longer used code   

this could probably be a separate doc for coding guidelines :sweat_smile:

File: docs/pull_request.md (hunk 16 above)
Comment (ouchadam) [reply docs/pull_request.md:37]: I wouldn't mind adding specific reviewers as additional reviewers :see_no_evil: I wonder if when a specific person is being added as a reviewer it's a sign that the area could benefit from more eyes for sharing the knowledge

File: docs/pull_request.md [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+In any case, it is better to explicitly declare in the description why the PR is a draft PR.
+
+Also, draft PR should not stay indefinitely in this state. It may be removed if it is the case and the submitter does not update it after a few days.
+
+##### PR Review Assignment
+
+We use automatic assignment for PR reviews. A PR is automatically routed by GitHub to a team member using the round robin algorithm. The process is the following:
+
+- The PR creator assigns the [element-android](https://github.com/orgs/vector-im/teams/element-android) team as a reviewer. They can skip this process and assign directly a specific member if they think they should take a look at it.
+- GitHub automatically assigns one reviewer. If the chosen reviewer is not available (holiday, etc.), remove them and manually add another reviewer.
```
Comment (manuroe) [inline docs/pull_request.md:38]:
```suggestion
- GitHub automatically assigns one reviewer. If the chosen reviewer is not available (holiday, etc.), remove them and set again the team, GitHub will select another reviewer.
```

File: docs/pull_request.md [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+
+##### PR description
+
+PR description should follow the PR template, and at least provide some context about the code change.
+
+##### File change
+
+1. Code should follow the guidelines
+2. Code should be formatted correctly
+3. XML attribute must be sorted
```
Comment (ouchadam) [inline docs/pull_request.md:132]: will the CI - lint/code quality script catch these? we might be able to delegate these points to the guidelines docs  (would be handy if the guidelines was a link)

File: docs/pull_request.md (hunk 8 above)
Comment (ouchadam) [reply docs/pull_request.md:187]: I wonder if it's worth introducing a label for _merge in next release cycle_ so that we could approve the PRs and provide more visibility for the anticipated merge window

File: docs/pull_request.md [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+4. Fork consideration. Will configuration of forks be easy? Some documentation may help in some cases.
+5. We are building long term products. "Quick and dirty" code must be avoided.
+6. The PR includes new tests for the added code, updated test for the existing code
+7. All PRs from external contributors **MUST** include a sign-off. It's in the checklist, and sometimes it's checked by the submitter, but there is actually no sign-off. In this case, ask nicely for a sign-off and request changes (do not approve the PR, even if everything else is fine).
+
+### Rules
+
+#### Check the form
+
+##### PR description
```
Comment (ouchadam) [inline docs/pull_request.md:124]: do you think we should have some conventions for the PR title as well?

File: docs/pull_request.md [hunk 24]
```diff
@@ -0,0 +1,208 @@
...
+It is up to the person who has merged the PR to go to the (closed) issue(s) and to add a comment to inform in which version the issue fix will be available. Use the current version of `develop` branch.
+
+> Closed in Element Android v1.x.y
+
+### Merge conflict
+
+It's up to the submitter to handle merge conflict. Sometimes, they can be fixed directly from GitHub, sometimes this is not possible. The branch can be rebased on `develop`, or the `develop` branch can be merged on the branch, it's up to the submitter to decide what is best.
+Keep in mind that Github Actions are not run in case of conflict.
+
+### When and who can merge PR
```
Comment (ouchadam) [inline docs/pull_request.md:185]: is it worth calling out how we merge? eg is there a preference for merging PRs as a squash/merge~ or the different scenarios where we might use one or the other (small fixes can be nice to squash)

**Review by ouchadam — APPROVED** [review]
:100: thanks for writing this up! I completely agree with the general sentiment, from my side I'm happy for this as a first pass (it's tricky to capture our entire process in one go)

File: docs/pull_request.md (hunk 18 above)
Comment (bmarty) [reply docs/pull_request.md:33]: I agree, but it impacts the whole toc. I can update that later

File: docs/pull_request.md (hunk 21 above)
Comment (bmarty) [reply docs/pull_request.md:38]: done

File: docs/pull_request.md (hunk 23 above)
Comment (bmarty) [reply docs/pull_request.md:124]: will add thanks

**Comment by bmarty** [comment]
I merge the PR, of course it still possible to update and improve the doc later!
